File: projects/tape/worker/src/bucketing.rs
//! Bucketing constants and pure helpers — Task 1.5.
//!
//! These live here (not in `bridge::messages`) because the bucket sizes
//! are **policy**, not **schema**. The wire shape (a `bucket_ts: i64`
//! and a `price_bucket: i64`) does not care what the bucket interval
//! is; the renderer and the aggregator do. Co-locating the constants
//! with the aggregator that uses them keeps schema evolution
//! (`messages.rs`) orthogonal to policy evolution (this file).
//!
//! Per PLAN.md § "Wow moment":
//!   - 1-minute time buckets, $5 price buckets on BTC-PERP.
//!
//! If a v2 multi-symbol expansion lands, replace the constants with a
//! per-symbol lookup (e.g. `bucket_size_for_symbol("ETH-PERP") = 0.5`).
//! The lookup stays in this module; the aggregator continues to call
//! the helpers without knowing about the new dimension.

/// Width of one time bucket (one footprint bar) in milliseconds.
/// PLAN.md pins 1 minute on BTC-PERP.
pub const TIME_BUCKET_MS: i64 = 60_000;

/// Width of one price bucket on BTC-PERP, in USD. PLAN.md § "Wow moment".
pub const PRICE_BUCKET_USD: f64 = 5.0;

/// Floor `ts_ms` to the start of its time bucket.
///
/// Pure function — no allocation, no I/O. Bench-friendly and trivially
/// testable.
#[inline]
#[must_use]
pub fn time_bucket(ts_ms: i64) -> i64 {
    if ts_ms >= 0 {
        (ts_ms / TIME_BUCKET_MS) * TIME_BUCKET_MS
    } else {
        // For ts < 0 (defensive — never expected on a real tick) we
        // floor toward negative infinity, matching JS Math.floor on the
        // TS side. ts_ms in this codebase is always a positive epoch
        // millisecond; the branch exists so a logic bug never trips a
        // signed-truncation surprise.
        ((ts_ms - TIME_BUCKET_MS + 1) / TIME_BUCKET_MS) * TIME_BUCKET_MS
    }
}

/// Floor `price` to its price-bucket index.
///
/// Returns `i64` because price buckets are dense integer ids in the
/// rendered chart (the X-axis is a bar index, the Y-axis is a price
/// bucket index). Multiply by `PRICE_BUCKET_USD` to recover the
/// lower-bound price of the bucket.
///
/// `f64::floor` is the canonical "floor toward negative infinity"
/// operation; `as i64` truncates toward zero, which would round the
/// wrong way for negative prices. BTC-PERP is never negative — but
/// EUR/USD differentials can be in v2, so we floor explicitly.
#[inline]
#[must_use]
#[allow(clippy::cast_possible_truncation)]
pub fn price_bucket(price: f64) -> i64 {
    (price / PRICE_BUCKET_USD).floor() as i64
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn time_bucket_floors_to_minute() {
        assert_eq!(time_bucket(0), 0);
        assert_eq!(time_bucket(59_999), 0);
        assert_eq!(time_bucket(60_000), 60_000);
        assert_eq!(time_bucket(60_001), 60_000);
        // 2026-05-30T12:34:56.789Z = 1_780_958_096_789 ms
        let ts: i64 = 1_780_958_096_789;
        let bucket = time_bucket(ts);
        assert_eq!(bucket % TIME_BUCKET_MS, 0);
        assert!(ts - bucket < TIME_BUCKET_MS);
    }

    #[test]
    fn price_bucket_floors_to_5_usd() {
        assert_eq!(price_bucket(0.0), 0);
        assert_eq!(price_bucket(4.99), 0);
        assert_eq!(price_bucket(5.0), 1);
        assert_eq!(price_bucket(7.5), 1);
        assert_eq!(price_bucket(10.0), 2);
        assert_eq!(price_bucket(71_234.5), 14_246);
        assert_eq!(price_bucket(71_235.0), 14_247);
    }
}

File: projects/tape/server/src/lib/aggregator/bucketing.ts
/**
 * Bucketing constants and pure helpers — Task 1.4 (TypeScript reference
 * implementation of the footprint-cell aggregator).
 *
 * This file is the TS MIRROR of the Rust worker's
 * `projects/tape/worker/src/bucketing.rs`. The two must stay
 * byte-for-byte equivalent in output because the conformance test
 * (Task 5.2) asserts the Rust port matches this reference on a recorded
 * tick dataset. Every change here MUST be mirrored in `bucketing.rs`
 * and vice versa — they are one logical module split across two
 * languages.
 *
 * **Why these constants live here, not in the WS / DB schema modules.**
 * The bucket sizes are POLICY, not SCHEMA. The wire shape (a
 * `bucketTs: number` and a `priceBucket: number`) does not care what
 * the bucket interval is; the renderer and the aggregator do. Keeping
 * the constants next to the aggregator that consumes them keeps schema
 * evolution orthogonal to policy evolution — identical reasoning to the
 * Rust side's module note.
 *
 * **v1 parameters (per PLAN.md § "Wow moment"):**
 *   - 1-minute time buckets (`TIME_BUCKET_MS = 60_000`).
 *   - $5 price buckets on BTC-PERP (`PRICE_BUCKET_USD = 5`).
 *
 * These are the values the Rust worker commits to
 * (`worker/src/bucketing.rs`). Per ADR-007 (single source of truth) the
 * synthesizer (`lib/ws/synthesizer.ts`) and the ingest snapshot-cache
 * path (`binance-ingestor.ts`) no longer hard-code `5` / `60_000`
 * inline — they import `priceBucket` / `timeBucket` / `TIME_BUCKET_MS`
 * from THIS module (Task 1.4c). The four-way duplication ADR-007 called
 * out is now collapsed to two irreducible literals (one per language —
 * here and in `bucketing.rs`) guarded by the Task 5.2 conformance
 * assertion. Do NOT re-inline `5` or `60_000` as bucketing math anywhere
 * else; that would defeat the guard, which only watches these two named
 * declarations.
 *
 * **v2 multi-symbol note.** When ETH-PERP / SOL-PERP land in v2, replace
 * the flat `PRICE_BUCKET_USD` constant with a per-symbol lookup
 * (`priceBucketSizeFor('ETH-PERP') = 0.5`). The lookup stays in this
 * module; callers keep calling the pure helpers without knowing about
 * the new dimension. Same migration path the Rust side documents.
 */

/**
 * Width of one time bucket (one footprint bar) in milliseconds.
 * PLAN.md pins 1 minute on BTC-PERP.
 */
export const TIME_BUCKET_MS = 60_000;

/**
 * Width of one price bucket on BTC-PERP, in USD. PLAN.md § "Wow moment".
 */
export const PRICE_BUCKET_USD = 5;

/**
 * Floor `tsMs` to the start of its time bucket.
 *
 * Pure function — no allocation, no I/O, no clock read. The input is a
 * tick timestamp; bar close is driven by tick timestamps crossing a
 * boundary, never by wall-clock, so replaying the same ticks always
 * yields identical buckets.
 *
 * For `tsMs >= 0` this is `floor(tsMs / TIME_BUCKET_MS) * TIME_BUCKET_MS`.
 * For `tsMs < 0` (defensive — never expected on a real epoch-ms tick) we
 * floor toward negative infinity to match the Rust side's signed-division
 * branch and JS `Math.floor` semantics. Real ticks in this codebase are
 * always positive epoch milliseconds; the branch exists so a logic bug
 * upstream never trips a truncation surprise that would diverge from the
 * Rust port.
 */
export function timeBucket(tsMs: number): number {
  return Math.floor(tsMs / TIME_BUCKET_MS) * TIME_BUCKET_MS;
}

/**
 * Floor `price` to its price-bucket INDEX.
 *
 * Returns a dense integer id (NOT a USD price). The rendered chart's
 * Y-axis is a price-bucket index; multiply by `PRICE_BUCKET_USD` to
 * recover the lower-bound price of the bucket. This INDEX convention is
 * the one the Rust worker (`price_bucket()` returns `i64`), the
 * synthesizer (`#pickPriceBucket` returns `floor(price / 5)`), and the
 * renderer's `priceToY` math all share. Emitting a USD-floored price
 * here instead would put every cell tens of thousands of pixels
 * off-viewport — a bug the synthesizer module documents having already
 * hit and fixed.
 *
 * `Math.floor` is "floor toward negative infinity"; BTC-PERP price is
 * never negative, but flooring explicitly keeps parity with the Rust
 * `f64::floor` (rather than `as i64` truncation toward zero) so a future
 * negative-price instrument in v2 buckets identically on both sides.
 */
export function priceBucket(price: number): number {
  return Math.floor(price / PRICE_BUCKET_USD);
}
